core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java

```diff
@@ -311,22 +312,37 @@ public void initialize() throws SailException {
 	}
 
 	@Override
-	public void shutDown() throws SailException {
+	public synchronized void shutDown() throws SailException {
 		if (shapesRepo != null) {
 			shapesRepo.shutDown();
 			shapesRepo = null;
 		}
 
 		if (executorService != null) {
 			executorService.shutdown();
+			boolean terminated = false;
+			try {
+				terminated = executorService.awaitTermination(200, TimeUnit.MILLISECONDS);
+			} catch (InterruptedException ignored) {
+			}
+
+			if (!terminated) {
+				executorService.shutdownNow();
+				logger.error("Shutdown ShaclSail while validation is still running.");
+			}
+
 		}
 
 		initialized.set(false);
+		executorService = null;
 		nodeShapes = Collections.emptyList();
 		super.shutDown();
 	}
 
-	public synchronized <T> Future<T> submitRunnableToExecutorService(Callable<T> runnable) {
+	synchronized <T> Future<T> submitRunnableToExecutorService(Callable<T> runnable) {
+		if (!initialized.get()) {
```

Comment:
this might be overkill

Reply:
Actually, we changed the contract on Sails in 3.0: it should auto-initialize when necessary, instead of throwing an exception. See #1223